lib/middleware/remote_user_auth_middleware.rb

```diff
@@ -0,0 +1,53 @@
+require 'securerandom'
+
+class RemoteUserAuthMiddleware
+
+  DefaultUserHeaderName = "Remote-User"
+  DefaultEmailHeaderName = "Remote-Email"
+  DefaultGroupsHeaderName = "Remote-Groups"
+  # if the header specified by user_header_name is provided in the
+  # request, treat the value of that header as a username, and log in as that
+  # user, creating if necessary.
+  # 
+  # This middleware can also be configured to detect headers containing the
+  # user's email, and comma separated group membership (along with a group name
+  # that signifies the created user should be an admin) during the user
+  # creation phase.
+  #
+  def initialize(app, user_header_name = nil, email_header_name = nil, groups_header_name = nil, admin_group = nil, warden_scope = :user)
+    @app = app
+    @user_header_name = (user_header_name || DefaultUserHeaderName).gsub("-", "_").upcase
+    @email_header_name = (email_header_name || DefaultEmailHeaderName).gsub("-", "_").upcase
+    @groups_header_name = (groups_header_name || DefaultGroupsHeaderName).gsub("-", "_").upcase
+    @admin_group = admin_group
+    @warden_scope = warden_scope
+  end
+
+  def call(env)
+    begin
+      username = env['HTTP_' + @user_header_name]
+      # do nothing if no remote-user is presented
+      if username
+        user = User.find_or_create_by!(username: username) do |u|
+          u.requires_no_invitation_code!
+
+          # needed for devise validations
+          u.password = SecureRandom.urlsafe_base64(24 + rand(8))
```

Comment:
Why `24 + rand(8)`? This looks like magic. -> Why not make it `32` and that should be good enough

Reply:
a fixed password length is a smaller search space for a variety of attacks. I have no preference here, switching to 32.